### ISSUE TITLE
Optimise "pass-through" cases

### ISSUE THREAD

**Issue by garyb** [purescript/purescript#1249]
The code generated for something like:
``` purescript
f (Prop e) = Prop e
```
Should be equivalent to:
``` purescript
f p = p
```
Even if it wouldn't have type checked had it been written that way (think of a case like this in a `Functor` implementation, for example).

**Comment by rightfold** [comment]
Started working on this: https://github.com/rightfold/purescript/commit/8d827f140630ff146c8be2ce85aad8f23b6e4fd1 :+1:

**Comment by garyb** [comment]
Yeah, it could definitely be applied there. We could do it in `CoreFn` too though, as the trick here is the type coercion, and we typecheck before producing the `CoreFn` AST.

**Comment by rightfold** [comment]
I'll take a look at that. Are optimisations already being performed on the `CoreFn` data structure? If so, in which modules does this take place? Or does this happen directly during the phase that generates `CoreFn`? The only optimisation modules I can find are JS-specific.

**Comment by rightfold** [comment]
Ok, I rewrote it to work on CoreFn instead of JS: https://github.com/rightfold/purescript/commit/0fb74c4a88e42517181d8dfbf074aa0567ec09f1

**Comment by garyb** [comment]
As you probably saw, we don't have an optimiser pass for `CoreFn` yet, but it was always the intention that we'd introduce something, as that's the reason `CoreFn` exists at all (that and the rewrites that we don't have yet). Your branch there looks pretty good to me, that's exactly what I would have done to introduce `CoreFn` optimisations. I've not looked at the actual optimisation too closely yet, if you open a PR we can review it properly.

Thanks for kicking this off!

**Comment by rightfold** [comment]
No problem. I enjoy it. :)
